src/models/app_data_json_format.rs

```diff
@@ -10,23 +10,40 @@ pub struct Referrer {
     pub version: String,
 }
 
-#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash, Default)]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash)]
+#[serde(untagged)]
+pub enum Quote {
+    V1(QuoteV1),
+    V2(QuoteV2),
+}
```

Comment:
I'm late to the party, but I think this needs to be tagged and not untagged (as @alfetopito suggested):

```rust
#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash)]
#[serde(tag = "version")]
pub enum Quote {
    #[serde(rename = "0.1.0")]
    V1(QuoteV1),
    #[serde(rename = "0.2.0")]
    V2(QuoteV2),
}
```

Otherwise, you may run into issues where a metadata file will:
- Serialize **without** the version - which is technically incorrect and may cause issues with compatibility for other consumers
- Deserialize to the wrong version if the same JSON payload is ambiguous (i.e. the JSON fits more than one version).

Reply:
I believe this was implemented in the expected way. Version is a string field that is passed (and parsed) with the content. The implementation here is sneakily avoiding the case of having awkward optional field logic by parsing a v2 with v1 parameters via the v1 type and still has version 0.2.0 in the version field. If we force the tags it is my impression that we will have to override the deserialization manually and implement the awkward optional field logic. Of course we can do this, but it's much messier.

Reply:
I had originally started with the "messier" alternative, but @MartinquaXD suggested this clever approach and I thought it was pretty sweet.

Reply:
Ah OK. So since we include the `version` as a `String`, we do serialize it correctly.

The part that bothers me is:
> Notice that although we have specified v2 for quote, we fall back on the v1 quote.

This means that:
```json
{
  "version": "2.0",
  "sellAmount": "1",
  "buyAmount": "1"
}
```

Serializes to a `V2` quote despite it being a `V1` quote, which is unexpected to me. Also, this lets you do stuff like:
```json
{
  "version": "not a version, but whatever - this is allowed",
  "slippageBips": "1"
}
```

Now, if you are querying the database for AppData entries with quotes v2 - you cannot rely on your query at all (because it might have a "0.2.0" label despite

> If we force the tags it is my impression that we will have to override the deserialization manually

The example I gave you works, and does not require any manual serialization code:

```rust
#[derive(Debug, Deserialize)]
#[serde(tag = "version")]
pub enum Quote {
    #[serde(rename = "0.1.0")]
    V1(QuoteV1),
    #[serde(rename = "0.2.0")]
    V2(QuoteV2),
}
```

See this [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bb24e7af959cd4d6e856252c75f02e45), where we try the following deserialization which is closer to what I would expect:
```rust
for json in [
    json!({ "version": "0.1.0", "sellAmount": "1", "buyAmount": "2" }),
    json!({ "version": "0.2.0", "slippageBips": "10" }),
    json!({ "version": "0.1.0", "slippageBips": "10" }), // invalid because the version is 0.1.0 despite the data being for version 0.2.0
    json!({ "version": "unknown", "slippageBips": "10" }), // invalid because the version is unknown
] {
    let _ = dbg!(serde_json::from_value::<Quote>(json));
}
```

Outputs:

```
[src/main.rs:40] serde_json::from_value::<Quote>(json) = Ok(
    V1(
        QuoteV1 {
            sell_amount: 1,
            buy_amount: 2,
        },
    ),
)
[src/main.rs:40] serde_json::from_value::<Quote>(json) = Ok(
    V2(
        QuoteV2 {
            slippage_bips: 10,
        },
    ),
)
[src/main.rs:40] serde_json::from_value::<Quote>(json) = Err(
    Error("missing field `sellAmount`", line: 0, column: 0),
)
[src/main.rs:40] serde_json::from_value::<Quote>(json) = Err(
    Error("unknown variant `unknown`, expected `0.1.0` or `0.2.0`", line: 0, column: 0),
)
```

Reply:
Ok cool! I will fix this in a follow up. Thanks for checking into it. I think I was just trying to avoid manually including this awkward optional field constraint logic. But doesn't this mean, if a quote comes in as version 0.2.0 but having v1 parameters that the parser will relabel it?

Reply:
> But doesn't this mean, if a quote comes in as version 0.2.0 but having v1 parameters that the parser will relabel it? I don't think we would want that because then the hashes wouldn't agree.

This is exactly my point, and what is happening now. The `version` field is essentially ignored and the deserialized data is being "relabel-ed" into a `QuoteVx` type despite what it says its version is.

Specifically, with my suggestion it does not get relabeled. I give an example of this above where deserializing this (quote with version `0.1.0` with v2 parameters) produces an error:
```json
{ "version": "0.1.0", "slippageBips": "10" }
```
```
[src/main.rs:40] serde_json::from_value::<Quote>(json) = Err(
    Error("missing field `sellAmount`", line: 0, column: 0),
)
```

Reply:
It is **NOT** being relabeled now. Why do you think that? it seems like I might as well just implement this the long way since it seems to be so difficult for everyone.